lib/livebook/runtime/node_pool.ex

```diff
@@ -0,0 +1,109 @@
+defmodule Livebook.Runtime.NodePool do
+  use GenServer
+
+  @moduledoc false
+
+  # A pool for reusing child node names.
+  #
+  # `free_name` refers to the list of unused names.
+  # `generated_names` refers to the list of names ever generated.
+  #
+  # `buffer_time` refers to time the pool waits before
+  # adding a name to `pool`,  which by default is 1 minute.
+
+  @default_time 60_000
+
+  # Client interface
+
+  @doc """
+  Starts the GenServer from a Supervision tree
+
+  ## Options
+
+    - `:name` - The name the NodePool is locally registered as. By default, it is `Livebook.Runtime.NodePool`
+    - `:buffer_time` - The time that is spent before a disconnected node's name is  added to pool. The default is 1 minute.
+  """
+  def start_link(opts) do
+    name = opts[:name] || __MODULE__
+    buffer_time = opts[:buffer_time] || @default_time
+
+    GenServer.start_link(
+      __MODULE__,
+      %{buffer_time: buffer_time},
+      name: name
+    )
+  end
+
+  @doc """
+  Retuns a node name.
+
+  Generates a new name if pool is empty, or takes one from pool.
+  """
+  def get_name(server \\ __MODULE__, basename) do
+    GenServer.call(server, {:get_name, basename})
+  end
+
+  # Server side code
+
+  @impl GenServer
+  def init(opts) do
+    :net_kernel.monitor_nodes(true, [{:node_type, :all}])
+
+    {:ok,
+     %{buffer_time: opts.buffer_time, generated_names: MapSet.new(), free_names: MapSet.new()}}
+  end
+
+  @impl GenServer
+  def handle_call({:get_name, basename}, _, state) do
+    {name, new_state} = name(state, basename)
+    {:reply, name, new_state}
+  end
+
+  @impl GenServer
+  def handle_info({:nodedown, node, _info}, state) do
+    {:ok, _} = :timer.send_after(state.buffer_time, __MODULE__, {:add_node, node})
+    {:noreply, state}
+  end
+
+  @impl GenServer
+  def handle_info({:nodeup, _node, _info}, state) do
+    # We don't want our mailbox full of node up messages
+    {:noreply, state}
```

Comment:
The messages need to be processed this way or another, so it's not like they would stay in the mailbox otherwise. If we didn't handle the message it would just result in function match error :)

```suggestion
    {:noreply, state}
```